File: desktop_client/api_client.py
"""
API Client for Leitor de Simulados Server
Handles all HTTP communication with the backend
"""
import requests
from pathlib import Path
from typing import Optional


class APIClient:
    """HTTP client for communicating with the Leitor de Simulados server"""
    
    def __init__(self, base_url: str = "http://localhost:8000"):
        self.base_url = base_url
        self.timeout = 30
    
    def _url(self, path: str) -> str:
        """Build full URL"""
        return f"{self.base_url.rstrip('/')}{path}"
    
    def _handle_response(self, response: requests.Response) -> dict:
        """Handle API response"""
        if response.status_code >= 400:
            try:
                error = response.json().get("detail", "Unknown error")
            except:
                error = response.text
            raise APIError(f"HTTP {response.status_code}: {error}")
        return response.json()
    
    # =========================================================================
    # Session endpoints
    # =========================================================================
    
    def create_session(self, test_type: str) -> dict:
        """Create a new processing session"""
        response = requests.post(
            self._url("/api/session/create"),
            data={"test_type": test_type},
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    def get_session(self, session_id: str) -> dict:
        """Get session information"""
        response = requests.get(
            self._url(f"/api/session/{session_id}"),
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    def delete_session(self, session_id: str) -> dict:
        """Delete a session"""
        response = requests.delete(
            self._url(f"/api/session/{session_id}"),
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    # =========================================================================
    # Model endpoints
    # =========================================================================
    
    def get_models(self) -> list:
        """Get available detection models"""
        response = requests.get(
            self._url("/api/models"),
            timeout=self.timeout
        )
        result = self._handle_response(response)
        return result.get("models", [])
    
    # =========================================================================
    # Image endpoints
    # =========================================================================
    
    def upload_images(self, session_id: str, file_paths: list) -> dict:
        """Upload images to a session"""
        files = []
        for path in file_paths:
            path = Path(path)
            if path.exists():
                files.append(
                    ("files", (path.name, open(path, "rb"), "image/jpeg"))
                )
        
        try:
            response = requests.post(
                self._url(f"/api/upload/{session_id}"),
                files=files,
                timeout=60  # Longer timeout for uploads
            )
            return self._handle_response(response)
        finally:
            # Close file handles
            for _, file_tuple in files:
                file_tuple[1].close()
    
    def get_images(self, session_id: str) -> dict:
        """Get list of images in a session"""
        response = requests.get(
            self._url(f"/api/images/{session_id}"),
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    def get_image(self, session_id: str, image_id: str, with_detections: bool = False) -> dict:
        """Get a specific image"""
        response = requests.get(
            self._url(f"/api/image/{session_id}/{image_id}"),
            params={"with_detections": with_detections},
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    # =========================================================================
    # Processing endpoints
    # =========================================================================
    
    def process_image(
        self,
        session_id: str,
        image_id: str,
        fs_model: str,
        ss_model: str,
        fs_threshold: float = 0.5,
        ss_threshold: float = 0.5
    ) -> dict:
        """Process a single image"""
        response = requests.post(
            self._url(f"/api/process/{session_id}/{image_id}"),
            data={
                "fs_model": fs_model,
                "ss_model": ss_model,
                "fs_threshold": fs_threshold,
                "ss_threshold": ss_threshold
            },
            timeout=120  # Long timeout for processing
        )
        return self._handle_response(response)
    
    def process_all_images(
        self,
        session_id: str,
        fs_model: str,
        ss_model: str,
        fs_threshold: float = 0.5,
        ss_threshold: float = 0.5
    ) -> dict:
        """Process all images in a session"""
        response = requests.post(
            self._url(f"/api/process-all/{session_id}"),
            data={
                "fs_model": fs_model,
                "ss_model": ss_model,
                "fs_threshold": fs_threshold,
                "ss_threshold": ss_threshold
            },
            timeout=600  # Very long timeout for batch processing
        )
        return self._handle_response(response)
    
    # =========================================================================
    # Report endpoints
    # =========================================================================
    
    def get_report(self, session_id: str, image_id: str) -> dict:
        """Get report for a processed image"""
        response = requests.get(
            self._url(f"/api/report/{session_id}/{image_id}"),
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    def get_all_reports(self, session_id: str) -> dict:
        """Get all reports for a session"""
        response = requests.get(
            self._url(f"/api/reports/{session_id}"),
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    def update_answer(
        self,
        session_id: str,
        image_id: str,
        question_number: int,
        answer: str
    ) -> dict:
        """Update an answer in the report"""
        response = requests.post(
            self._url(f"/api/update-answer/{session_id}/{image_id}"),
            data={
                "question_number": question_number,
                "answer": answer
            },
            timeout=self.timeout
        )
        return self._handle_response(response)
    
    def export_reports(self, session_id: str, format: str = "json") -> dict:
        """Export all reports"""
        response = requests.get(
            self._url(f"/api/export/{session_id}"),
            params={"format": format},
            timeout=self.timeout
        )
        return self._handle_response(response)


class APIError(Exception):
    """API communication error"""
    pass

File: cpp_client/src/api_client.cpp
#include "api_client.h"
#include <curl/curl.h>
#include <sstream>
#include <fstream>
#include <stdexcept>

namespace leitor {

// CURL write callback
static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* userp) {
    size_t total = size * nmemb;
    userp->append((char*)contents, total);
    return total;
}

APIClient::APIClient(const std::string& base_url) : base_url_(base_url) {
    curl_global_init(CURL_GLOBAL_ALL);
}

APIClient::~APIClient() {
    curl_global_cleanup();
}

bool APIClient::test_connection() {
    try {
        auto result = http_get("/api/models");
        return result.contains("models");
    } catch (...) {
        return false;
    }
}

std::string APIClient::create_session(const std::string& test_type) {
    auto result = http_post_form("/api/session/create", {{"test_type", test_type}});
    return result["session_id"].get<std::string>();
}

json APIClient::get_session(const std::string& session_id) {
    return http_get("/api/session/" + session_id);
}

bool APIClient::delete_session(const std::string& session_id) {
    try {
        http_delete("/api/session/" + session_id);
        return true;
    } catch (...) {
        return false;
    }
}

std::vector<ModelInfo> APIClient::get_models() {
    auto result = http_get("/api/models");
    std::vector<ModelInfo> models;
    
    for (const auto& m : result["models"]) {
        ModelInfo info;
        info.name = m["name"].get<std::string>();
        info.model_type = m["model_type"].get<std::string>();
        info.target_stage = m["target_stage"].get<std::string>();
        info.rel_path = m["rel_path"].get<std::string>();
        models.push_back(info);
    }
    
    return models;
}

json APIClient::upload_images(const std::string& session_id, const std::vector<std::string>& file_paths) {
    return http_post_files("/api/upload/" + session_id, file_paths);
}

std::vector<ImageInfo> APIClient::get_images(const std::string& session_id) {
    auto result = http_get("/api/images/" + session_id);
    std::vector<ImageInfo> images;
    
    for (const auto& img : result["images"]) {
        ImageInfo info;
        info.id = img["id"].get<std::string>();
        info.filename = img["filename"].get<std::string>();
        info.processed = img["processed"].get<bool>();
        info.has_report = img["has_report"].get<bool>();
        images.push_back(info);
    }
    
    return images;
}

std::vector<unsigned char> APIClient::get_image_data(const std::string& session_id, 
                                                      const std::string& image_id, 
                                                      bool with_detections) {
    std::string endpoint = "/api/image/" + session_id + "/" + image_id;
    if (with_detections) {
        endpoint += "?with_detections=true";
    }
    
    auto result = http_get(endpoint);
    std::string base64_data = result["image"].get<std::string>();
    
    // Decode base64
    // Using simple base64 decode (see image_loader.cpp)
    extern std::vector<unsigned char> base64_decode(const std::string&);
    return base64_decode(base64_data);
}

ProcessingResult APIClient::process_image(const std::string& session_id,
                                          const std::string& image_id,
                                          const std::string& fs_model,
                                          const std::string& ss_model,
                                          float fs_threshold,
                                          float ss_threshold) {
    auto result = http_post_form("/api/process/" + session_id + "/" + image_id, {
        {"fs_model", fs_model},
        {"ss_model", ss_model},
        {"fs_threshold", std::to_string(fs_threshold)},
        {"ss_threshold", std::to_string(ss_threshold)}
    });
    
    ProcessingResult pr;
    pr.status = result["status"].get<std::string>();
    pr.image_id = result["image_id"].get<std::string>();
    pr.detections_count = result["detections_count"].get<int>();
    
    if (result.contains("report") && !result["report"].is_null()) {
        Report report;
        report.owner_cpf = result["report"]["owner_cpf"].get<std::string>();
        report.test_type = result["report"]["test_type"].get<std::string>();
        
        for (const auto& q : result["report"]["questions"]) {
            QuestionAnswer qa;
            qa.number = q["number"].get<int>();
            qa.answer = q["answer"].get<std::string>();
            qa.updated = q["updated"].get<bool>();
            report.questions.push_back(qa);
        }
        pr.report = report;
    }
    
    return pr;
}

std::vector<ProcessingResult> APIClient::process_all_images(const std::string& session_id,
                                                             const std::string& fs_model,
                                                             const std::string& ss_model,
                                                             float fs_threshold,
                                                             float ss_threshold) {
    auto result = http_post_form("/api/process-all/" + session_id, {
        {"fs_model", fs_model},
        {"ss_model", ss_model},
        {"fs_threshold", std::to_string(fs_threshold)},
        {"ss_threshold", std::to_string(ss_threshold)}
    });
    
    std::vector<ProcessingResult> results;
    for (const auto& r : result["results"]) {
        ProcessingResult pr;
        pr.status = r["status"].get<std::string>();
        pr.image_id = r["image_id"].get<std::string>();
        pr.detections_count = r.value("detections_count", 0);
        results.push_back(pr);
    }
    
    return results;
}

Report APIClient::get_report(const std::string& session_id, const std::string& image_id) {
    auto result = http_get("/api/report/" + session_id + "/" + image_id);
    
    Report report;
    report.owner_cpf = result["report"]["owner_cpf"].get<std::string>();
    report.test_type = result["report"]["test_type"].get<std::string>();
    
    for (const auto& q : result["report"]["questions"]) {
        QuestionAnswer qa;
        qa.number = q["number"].get<int>();
        qa.answer = q["answer"].get<std::string>();
        qa.updated = q["updated"].get<bool>();
        report.questions.push_back(qa);
    }
    
    return report;
}

std::vector<Report> APIClient::get_all_reports(const std::string& session_id) {
    auto result = http_get("/api/reports/" + session_id);
    std::vector<Report> reports;
    
    for (const auto& r : result["reports"]) {
        Report report;
        report.owner_cpf = r["report"]["owner_cpf"].get<std::string>();
        report.test_type = r["report"]["test_type"].get<std::string>();
        
        for (const auto& q : r["report"]["questions"]) {
            QuestionAnswer qa;
            qa.number = q["number"].get<int>();
            qa.answer = q["answer"].get<std::string>();
            qa.updated = q["updated"].get<bool>();
            report.questions.push_back(qa);
        }
        reports.push_back(report);
    }
    
    return reports;
}

bool APIClient::update_answer(const std::string& session_id, 
                               const std::string& image_id,
                               int question_number, 
                               const std::string& answer) {
    try {
        http_post_form("/api/update-answer/" + session_id + "/" + image_id, {
            {"question_number", std::to_string(question_number)},
            {"answer", answer}
        });
        return true;
    } catch (...) {
        return false;
    }
}

std::string APIClient::export_reports(const std::string& session_id, const std::string& format) {
    auto result = http_get("/api/export/" + session_id + "?format=" + format);
    return result.dump(2);
}

// HTTP implementation using libcurl
json APIClient::http_get(const std::string& endpoint) {
    CURL* curl = curl_easy_init();
    std::string response;
    
    if (curl) {
        std::string url = base_url_ + endpoint;
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
        
        CURLcode res = curl_easy_perform(curl);
        long http_code = 0;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
        curl_easy_cleanup(curl);
        
        if (res != CURLE_OK) {
            last_error_ = curl_easy_strerror(res);
            throw std::runtime_error(last_error_);
        }
        
        if (http_code >= 400) {
            last_error_ = "HTTP " + std::to_string(http_code);
            throw std::runtime_error(last_error_);
        }
    }
    
    return json::parse(response);
}

json APIClient::http_post(const std::string& endpoint, const std::string& data) {
    CURL* curl = curl_easy_init();
    std::string response;
    
    if (curl) {
        std::string url = base_url_ + endpoint;
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 120L);
        
        CURLcode res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
        
        if (res != CURLE_OK) {
            last_error_ = curl_easy_strerror(res);
            throw std::runtime_error(last_error_);
        }
    }
    
    return json::parse(response);
}

json APIClient::http_post_form(const std::string& endpoint, 
                                const std::vector<std::pair<std::string, std::string>>& fields) {
    CURL* curl = curl_easy_init();
    std::string response;
    
    if (curl) {
        std::string url = base_url_ + endpoint;
        
        // Build form data
        curl_mime* form = curl_mime_init(curl);
        for (const auto& field : fields) {
            curl_mimepart* part = curl_mime_addpart(form);
            curl_mime_name(part, field.first.c_str());
            curl_mime_data(part, field.second.c_str(), CURL_ZERO_TERMINATED);
        }
        
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_MIMEPOST, form);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 120L);
        
        CURLcode res = curl_easy_perform(curl);
        long http_code = 0;
        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
        
        curl_mime_free(form);
        curl_easy_cleanup(curl);
        
        if (res != CURLE_OK) {
            last_error_ = curl_easy_strerror(res);
            throw std::runtime_error(last_error_);
        }
        
        if (http_code >= 400) {
            last_error_ = "HTTP " + std::to_string(http_code) + ": " + response;
            throw std::runtime_error(last_error_);
        }
    }
    
    return json::parse(response);
}

json APIClient::http_post_files(const std::string& endpoint, const std::vector<std::string>& file_paths) {
    CURL* curl = curl_easy_init();
    std::string response;
    
    if (curl) {
        std::string url = base_url_ + endpoint;
        
        curl_mime* form = curl_mime_init(curl);
        for (const auto& path : file_paths) {
            curl_mimepart* part = curl_mime_addpart(form);
            curl_mime_name(part, "files");
            curl_mime_filedata(part, path.c_str());
        }
        
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_MIMEPOST, form);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 300L);  // Longer timeout for uploads
        
        CURLcode res = curl_easy_perform(curl);
        curl_mime_free(form);
        curl_easy_cleanup(curl);
        
        if (res != CURLE_OK) {
            last_error_ = curl_easy_strerror(res);
            throw std::runtime_error(last_error_);
        }
    }
    
    return json::parse(response);
}

json APIClient::http_delete(const std::string& endpoint) {
    CURL* curl = curl_easy_init();
    std::string response;
    
    if (curl) {
        std::string url = base_url_ + endpoint;
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "DELETE");
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
        
        CURLcode res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
        
        if (res != CURLE_OK) {
            last_error_ = curl_easy_strerror(res);
            throw std::runtime_error(last_error_);
        }
    }
    
    return json::parse(response);
}

std::string APIClient::url_encode(const std::string& str) {
    CURL* curl = curl_easy_init();
    char* encoded = curl_easy_escape(curl, str.c_str(), str.length());
    std::string result(encoded);
    curl_free(encoded);
    curl_easy_cleanup(curl);
    return result;
}

} // namespace leitor
